sources/scraping/runner.py

```diff
@@ -0,0 +1,133 @@
+import threading
+import typing as t
+
+import dlt
+from dlt.common import logger
+from pydispatch import dispatcher
+
+from scrapy import signals, Item, Spider  # type: ignore
+from scrapy.crawler import CrawlerProcess  # type: ignore
+
+from .types import AnyDict, Runnable, P
+from .queue import ScrapingQueue
+
+
+class ScrapyRunner(Runnable):
+    def __init__(
+        self,
+        spider: t.Type[Spider],
+        start_urls: t.List[str],
+        settings: AnyDict,
+        on_item_scraped: t.Callable[[Item], None],
+        on_engine_stopped: t.Callable[[], None],
+    ) -> None:
+        self.spider = spider
+        self.settings = settings
+        self.start_urls = start_urls
+
+        # We want to receive on_item_scraped callback from
+        # outside so we don't have to know about any queue instance.
+        self.on_item_scraped = on_item_scraped
+        self.on_engien_stopped = on_engine_stopped
+
+    def run(self, *args: P.args, **kwargs: P.kwargs) -> t.Any:
+        crawler = CrawlerProcess(settings=self.settings)
+        crawler.crawl(
+            self.spider,
+            name="scraping_spider",
+            start_urls=self.start_urls,
+            **kwargs,
+        )
+
+        # Setup signals
+        dispatcher.connect(self.on_item_scraped, signals.item_scraped)
+        dispatcher.connect(self.on_engien_stopped, signals.engine_stopped)
+
+        try:
+            crawler.start()
+        except Exception:
+            logger.error("Was unable to start crawling process")
+            self.on_engien_stopped()
+            raise
+
+
+class PipelineRunner(Runnable):
+    def __init__(
+        self,
+        pipeline: dlt.Pipeline,
+        queue: ScrapingQueue,
+    ) -> None:
+        self.pipeline = pipeline
+        self.queue = queue
+
+    def run(  # type: ignore[override]
+        self,
+        data: t.Any,
+        *args: P.args,
+        **kwargs: P.kwargs,
+    ) -> threading.Thread:
+        """You can use all regular dlt.pipeline.run() arguments
+
+        ```
+        destination: TDestinationReferenceArg = None,
+        staging: TDestinationReferenceArg = None,
+        dataset_name: str = None,
+        credentials: Any = None,
+        table_name: str = None,
+        write_disposition: TWriteDisposition = None,
+        columns: TAnySchemaColumns = None,
+        primary_key: TColumnNames = None,
+        schema: Schema = None,
+        loader_file_format: TLoaderFileFormat = None
+        ```
+        """
+
+        def run() -> None:
+            try:
+                self.pipeline.run(data, **kwargs)
+            except Exception:
+                logger.error("Error during pipeline.run call, closing the queue")
+                self.queue.close()
+                raise
```

Comment:
just propagate the exception, raising CloseSpider won't have any effect on scrapy because we are running the pipeline in a separate thread.